File: mcp-code-intelligence-nodejs/src/memory/migrations-v4.ts
/**
 * V4 migration runner — adds agent_name column to knowledge_entries.
 * Idempotent — safe to call multiple times.
 */

import Database from 'better-sqlite3';
import { SCHEMA_V4_AGENT_NAME_ALTER, SCHEMA_V4_AGENT_NAME_INDEX } from './schema-v4.js';

/** Run all V4 migrations (idempotent). */
export function runV4Migrations(db: Database.Database): void {
  runAlterStatements(db, SCHEMA_V4_AGENT_NAME_ALTER);
  safeExec(db, SCHEMA_V4_AGENT_NAME_INDEX);
}

/** Run ALTER statements one by one, ignoring "duplicate column" errors. */
function runAlterStatements(db: Database.Database, stmts: string[]): void {
  for (const sql of stmts) {
    try {
      db.exec(sql);
    } catch (err: unknown) {
      const msg = (err as Error).message ?? '';
      if (!msg.includes('duplicate column')) throw err;
    }
  }
}

/** Execute multi-statement SQL, ignoring "already exists" errors. */
function safeExec(db: Database.Database, sql: string): void {
  try {
    db.exec(sql);
  } catch (err: unknown) {
    const msg = (err as Error).message ?? '';
    if (!msg.includes('already exists')) throw err;
  }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/migrations_v4.py
"""V4 migration runner — adds agent_name column to knowledge_entries."""

import logging
import sqlite3

from .schema_v4 import SCHEMA_V4_AGENT_NAME_ALTER, SCHEMA_V4_AGENT_NAME_INDEX

logger = logging.getLogger(__name__)


def run_v4_migrations(conn: sqlite3.Connection) -> None:
    """Run all V4 migrations (idempotent — safe to call multiple times)."""
    _run_alter_statements(conn, SCHEMA_V4_AGENT_NAME_ALTER)
    _safe_exec(conn, SCHEMA_V4_AGENT_NAME_INDEX)
    conn.commit()
    logger.info("V4 migrations applied successfully")


def _run_alter_statements(conn: sqlite3.Connection, stmts: list[str]) -> None:
    """Run ALTER statements one by one, ignoring duplicate column errors."""
    cursor = conn.cursor()
    for sql in stmts:
        try:
            cursor.execute(sql)
        except sqlite3.OperationalError as e:
            if "duplicate column" not in str(e):
                raise


def _safe_exec(conn: sqlite3.Connection, sql: str) -> None:
    """Execute multi-statement SQL, ignoring already exists errors."""
    cursor = conn.cursor()
    for stmt in sql.strip().split(";"):
        stmt = stmt.strip()
        if not stmt:
            continue
        try:
            cursor.execute(stmt)
        except sqlite3.OperationalError as e:
            if "already exists" not in str(e):
                raise
